File: mirror/tier2-php/showcase/perf_benchmark.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Optimization
// Module: perf_benchmark.php

function measure_time(callable $func, int $iterations = 1000): array {
    $start = microtime(true);
    
    for ($i = 0; $i < $iterations; $i++) {
        $func();
    }
    
    $elapsed = (microtime(true) - $start) * 1000;
    
    return [
        'elapsed_ms' => round($elapsed, 3),
        'iterations' => $iterations,
        'per_call_us' => round($elapsed * 1000 / $iterations, 2),
    ];
}

function simple_operation(int $x): int {
    return $x * 2;
}

function complex_operation(int $x): int {
    $result = 0;
    for ($i = 0; $i < 10; $i++) {
        $result += $x * $i;
    }
    return $result;
}

function optimized_operation(int $x): int {
    return intval($x * 10 * 9 / 2);
}

$simpleBench = measure_time(fn() => simple_operation(100), 10000);
$complexBench = measure_time(fn() => complex_operation(100), 1000);
$optimizedBench = measure_time(fn() => optimized_operation(100), 10000);

$result = [
    'simple_elapsed' => $simpleBench['elapsed_ms'],
    'complex_elapsed' => $complexBench['elapsed_ms'],
    'optimized_elapsed' => $optimizedBench['elapsed_ms'],
    'benchmarking_active' => true,
];

echo json_encode($result);
?>

File: mirror/tier2-python/showcase/perf_benchmark.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Optimization
Module: perf_benchmark.py
Purpose: Demonstrate performance benchmarking
"""

import time
from typing import Callable, Any


def measure_time(func: Callable, *args, iterations: int = 1000) -> dict:
    """Measure function execution time"""
    start = time.perf_counter()
    
    for _ in range(iterations):
        func(*args)
    
    end = time.perf_counter()
    elapsed = (end - start) * 1000  # Convert to ms
    
    return {
        "elapsed_ms": round(elapsed, 3),
        "iterations": iterations,
        "per_call_us": round(elapsed * 1000 / iterations, 2),
    }


def simple_operation(x: int) -> int:
    """Simple operation"""
    return x * 2


def complex_operation(x: int) -> int:
    """Complex operation"""
    result = 0
    for i in range(10):
        result += x * i
    return result


def optimized_operation(x: int) -> int:
    """Optimized operation"""
    return x * 10 * 9 // 2  # Closed form


simple_bench = measure_time(simple_operation, 100, iterations=10000)
complex_bench = measure_time(complex_operation, 100, iterations=1000)
optimized_bench = measure_time(optimized_operation, 100, iterations=10000)

result = {
    "simple_elapsed": simple_bench["elapsed_ms"],
    "complex_elapsed": complex_bench["elapsed_ms"],
    "optimized_elapsed": optimized_bench["elapsed_ms"],
    "simple_per_call": simple_bench["per_call_us"],
    "benchmarking_active": True,
}

if __name__ == "__main__":
    print(result)
